sessions/test_js_challenge.py

```diff
@@ -0,0 +1,306 @@
+            'addr' : "${tempesta_ip}",
+            'port' : '80',
+        },
+        {
+            'id' : 'client-2',
+            'type' : 'deproxy',
+            'addr' : "${tempesta_ip}",
+            'port' : '80',
+        },
+        {
+            'id' : 'client-3',
+            'type' : 'deproxy',
+            'addr' : "${tempesta_ip}",
+            'port' : '80',
+        }
+    ]
+
+    def wait_all_connections(self, tmt=1):
+        sids = self.get_servers_id()
+        for sid in sids:
+            srv = self.get_server(sid)
+            if not srv.wait_for_connections(timeout=tmt):
+                return False
+        return True
+
+    def client_send_req(self, client, req):
+        curr_responses = len(client.responses)
+        client.make_requests(req)
+        client.wait_for_response(timeout=1)
+        self.assertEqual(curr_responses + 1, len(client.responses))
+
+        return client.responses[-1]
+
+    def client_expect_block(self, client, req):
+        curr_responses = len(client.responses)
+        client.make_requests(req)
+        client.wait_for_response(timeout=1)
+        self.assertEqual(curr_responses, len(client.responses))
+        self.assertTrue(client.connection_is_closed())
+
+    def prepare_js_templates(self):
+        srcdir = tf_cfg.cfg.get('Tempesta', 'srcdir')
+        workdir = tf_cfg.cfg.get('Tempesta', 'workdir')
+        template = "%s/etc/js_challenge.tpl" % srcdir
+        js_code = "%s/etc/js_challenge.js.tpl" % srcdir
+        remote.tempesta.copy_file_to_node(js_code, workdir)
+        remote.tempesta.copy_file_to_node(template, "%s/js1.tpl" % workdir)
+        remote.tempesta.copy_file_to_node(template, "%s/js2.tpl" % workdir)
+        remote.tempesta.copy_file_to_node(template, "%s/js3.tpl" % workdir)
+
+    def start_all(self):
+        self.prepare_js_templates()
+        self.start_all_servers()
+        self.start_tempesta()
+        self.start_all_clients()
+        self.deproxy_manager.start()
+        self.assertTrue(self.wait_all_connections(1))
+
+    def test_get_challenge(self):
+        self.start_all()
+        client = self.get_client('client-1')
+
+        # Client can accept JS code in responses.
+        req = ("GET / HTTP/1.1\r\n"
+               "Host: vh1.com\r\n"
+               "Accept: text/html\r\n"
+               "\r\n")
+        resp = self.client_send_req(client, req)
+        self.assertEqual(resp.status, '503',
+                         "Unexpected response status code")
+        self.assertIsNotNone(resp.headers.get('Set-Cookie', None),
+                             "Set-Cookie header is missing in the response")
+        self.assertIsNotNone(resp.headers.get('Set-Cookie', None),
+                             "Set-Cookie header is missing in the response")
+        match = re.search(r'(location\.reload)', resp.body)
+        self.assertIsNotNone(match,
+                             "Cant extract value from Set-Cookie header")
+
```

Comment:
In this test we trying to check that the JS challenge is sent to client if it can accept it, otherwise the connection is closed. Not all requests can be challenged - only GET with proper accept headers. I will add the comment.